### PR TITLE
Handle TupleXXL in match analysis

### PR DESCRIPTION
**Pull request by dwijnand**
There's a number of problems with the match analysis of TupleXXL.
Of course, they manifest as (in)exhaustivity and (un)reachability
warnings.

Reachability suffered by the problem that a large generic tuple
scrutinee type wasn't considered extractable by the TupleXXL extractor
that Typer associates the extractor pattern with.  That was solved by
special handling in SpaceEngine's isSubType.

Exhaustivity suffered by a variety of problems, again stemming from the
disconnect between the TupleXXL pattern type and the large generic tuple
scrutinee (or component) type.  That was solved by special handling in
exhaustivityCheckable to ignore large generic tuple scrutinees.

That then highlighted an irrefutable failure (checkIrrefutable), which
also needed to be taught that extra large generic tuples do conform to
TupleXXL extractors type, after which SpaceEngine isIrrefutable needed
special handling to consider TupleXXL irrefutable.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala [hunk 1]
```diff
@@ -836,7 +835,7 @@ object SpaceEngine {
     def isCheckable(tp: Type): Boolean =
       val tpw = tp.widen.dealias
       val classSym = tpw.classSymbol
-      classSym.is(Sealed) ||
+      classSym.is(Sealed) && !tpw.isLargeGenericTuple || // exclude large generic tuples from exhaustivity
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/patmat/Space.scala:838]: Could the comment highlight that this is a limitation and what would need to be done to lift this limitation?

**Review by smarter — CHANGES_REQUESTED** [review]
If I understand correctly this turns off exhaustivity checking for TupleXXL but keeps reachability checking, right? In that case, could you add a test case where we detect an unreachable case?

**Review by smarter — APPROVED** [review]

**Comment by dwijnand** [comment]
I went back down the rabbit hole trying to get a sense of what changes are required, and there are so many different parts and I can't tell which behaviour needs changing, particularly because I rarely spend time on unapplySeq fixes.

**Comment by odersky** [comment]
Great that this could be fixed!

**Pull request merged**